File: Pin.py
from enum import Enum

from EnumClass import ALL_CLOCK_EDGES, ALL_TIMING_MODES, EnumClockEdge, EnumPinType, EnumTimingMode

class Pin:
    def __init__(self, name, pin_type=None):
        self.name = name
        self.net = None
        self.type = pin_type  # 避免与Python的type关键字冲突
        self.capacitance = {EnumTimingMode.MAX: {EnumClockEdge.RISING: 0.0, EnumClockEdge.FALLING: 0.0}, 
                            EnumTimingMode.MIN: {EnumClockEdge.RISING: 0.0, EnumClockEdge.FALLING: 0.0}}
        self.slew = {EnumTimingMode.MAX: {EnumClockEdge.RISING: None, EnumClockEdge.FALLING: None}, 
                     EnumTimingMode.MIN: {EnumClockEdge.RISING: None, EnumClockEdge.FALLING: None}}
        self.arrival_time = {EnumTimingMode.MAX: {EnumClockEdge.RISING: None, EnumClockEdge.FALLING: None}, 
                      EnumTimingMode.MIN: {EnumClockEdge.RISING: None, EnumClockEdge.FALLING: None}}
        self.predecessor = {EnumTimingMode.MAX: {EnumClockEdge.RISING: None, EnumClockEdge.FALLING: None},
                            EnumTimingMode.MIN: {EnumClockEdge.RISING: None, EnumClockEdge.FALLING: None}}
        self.arcs = None # deprecated
        self.fanin = []
        self.fanout = []
    
    def __repr__(self):
        return f"Pin({self.name}, type={self.type})"

    def get_slew(self, timing_mode, clock_edge):
        """获取Pin的slew值"""
        assert timing_mode in ALL_TIMING_MODES
        assert clock_edge in ALL_CLOCK_EDGES
        return self.slew[timing_mode][clock_edge]

    def set_slew(self, timing_mode, clock_edge, slew_value):
        """更新Pin的slew值"""
        assert timing_mode in ALL_TIMING_MODES
        assert clock_edge in ALL_CLOCK_EDGES
        self.slew[timing_mode][clock_edge] = slew_value

    def get_capacitance(self, timing_mode, clock_edge):
        """获取Pin的capacitance值"""
        assert timing_mode in ALL_TIMING_MODES
        assert clock_edge in ALL_CLOCK_EDGES
        return self.capacitance[timing_mode][clock_edge]

    def set_capacitance(self, timing_mode, clock_edge, capacitance_value):
        """更新Pin的capacitance值"""
        assert timing_mode in ALL_TIMING_MODES
        assert clock_edge in ALL_CLOCK_EDGES
        self.capacitance[timing_mode][clock_edge] = capacitance_value

    def update_capacitance(self, timing_mode, clock_edge):
        total_capacitance = sum(arc.get_capacitance(timing_mode, clock_edge) for arc in self.fanout)
        self.set_capacitance(timing_mode, clock_edge, total_capacitance)

    def connect_to_net(self, net):
        """将Pin连接到Net，根据Pin类型确定是source还是sink"""
        assert net is not None
        self.net = net
        if self.type == EnumPinType.INPUT or self.type == EnumPinType.PRIMARY_OUTPUT:
            net.add_sink(self)
        elif self.type == EnumPinType.OUTPUT or self.type == EnumPinType.PRIMARY_INPUT:
            net.set_source(self)
        else:
            raise ValueError(f"Unknown pin type: {self.type} for pin {self.name}")

    def propagate_slew(self, timing_mode, from_clock_edge, to_clock_edge):
        """传播Pin的slew值到fanout的Arc和to_pin"""
        for arc in self.fanout:
            arc.propagate_slew(timing_mode, from_clock_edge, to_clock_edge)

    def propagate_delay(self, timing_mode, from_clock_edge, to_clock_edge):
        """传播Pin的arrival_time值到fanout的Arc和to_pin"""
        for arc in self.fanout:
            arc.propagate_delay(timing_mode, from_clock_edge, to_clock_edge)

    def update_predecessor(self, timing_mode, clock_edge, predecessor_arc):
        """更新Pin的predecessor"""
        self.predecessor[timing_mode][clock_edge] = predecessor_arc

    def update_slew(self, timing_mode, clock_edge, slew_value):
        """更新Pin的slew值"""
        old_slew = self.get_slew(timing_mode, clock_edge)
        if timing_mode == EnumTimingMode.MAX:
            if not old_slew or slew_value > old_slew:
                self.set_slew(timing_mode, clock_edge, slew_value)
        elif timing_mode == EnumTimingMode.MIN:
            if not old_slew or slew_value < old_slew:
                self.set_slew(timing_mode, clock_edge, slew_value)
        else:
            raise ValueError(f"Unknown timing mode: {timing_mode} for pin {self.name}")
    
    def get_arrival_time(self, timing_mode, clock_edge):
        """获取Pin的arrival_time值"""
        assert timing_mode in ALL_TIMING_MODES
        assert clock_edge in ALL_CLOCK_EDGES
        at = self.arrival_time[timing_mode][clock_edge]
        return at
    
    def set_arrival_time(self, timing_mode, clock_edge, arrival_time_value):
        """更新Pin的arrival_time值"""
        assert timing_mode in ALL_TIMING_MODES
        assert clock_edge in ALL_CLOCK_EDGES
        self.arrival_time[timing_mode][clock_edge] = arrival_time_value
    
    def update_arrival_time(self, timing_mode, clock_edge, arrival_time_value, from_arc):
        """更新Pin的arrival_time值"""
        assert timing_mode in ALL_TIMING_MODES
        assert clock_edge in ALL_CLOCK_EDGES
        old_at = self.get_arrival_time(timing_mode, clock_edge)
        if timing_mode == EnumTimingMode.MAX:
            if not old_at or arrival_time_value > old_at:
                self.set_arrival_time(timing_mode, clock_edge, arrival_time_value)
                self.update_predecessor(timing_mode, clock_edge, from_arc)
        elif timing_mode == EnumTimingMode.MIN:
            if not old_at or arrival_time_value < old_at:
                self.set_arrival_time(timing_mode, clock_edge, arrival_time_value)
                self.update_predecessor(timing_mode, clock_edge, from_arc)
        else:
            raise ValueError(f"Unknown timing mode: {timing_mode} for pin {self.name}")

    def propagate_arrival_time(self, timing_mode, from_clock_edge, to_clock_edge):
        """传播Pin的arrival_time值到fanout的Arc和to_pin"""
        for arc in self.fanout:
            arc.propagate_arrival_time(timing_mode, from_clock_edge, to_clock_edge)

class PinFactory:
    """Pin对象工厂，管理Pin的创建和查询"""
    def __init__(self):
        self._pins = {}
    
    def create_pin(self, name, pin_type=None):
        if name in self._pins:
            return self._pins[name]
        
        pin = Pin(name, pin_type)
        self._pins[name] = pin
        return pin
    
    def get_pin(self, name):
        return self._pins.get(name)
    
    def get_all_pins(self):
        return list(self._pins.values())

File: src/Pin.cc
//
// Created by wenz on 2/23/26.
//

#include "Pin.h"

#include <algorithm>
#include <ranges>

#include "Arc.h"
#include "Enum/EnumForeach.h"
#include "Enum/EnumPinType.h"
#include "Log.hh"

namespace mySTA {

opt_float_t Pin::get_slew(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge) const
{
  return _slew[+timing_mode][+clock_edge];
}

float_t Pin::get_capacitance(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge) const
{
  return _capacitance[+timing_mode][+clock_edge];
}

opt_float_t Pin::get_arrival_time(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge) const
{
  return _arrival_time[+timing_mode][+clock_edge];
}

opt_float_t Pin::get_request_arrival_time(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge) const
{
  return _request_arrival_time[+timing_mode][+clock_edge];
}

void Pin::update_slew(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge, const float_t new_slew)
{
  const auto old_slew{get_slew(timing_mode, clock_edge)};
  switch (timing_mode) {
    case EnumTimingMode::MAX:
      if (!old_slew || new_slew > *old_slew) {
        set_slew(timing_mode, clock_edge, new_slew);
      }
      break;
    case EnumTimingMode::MIN:
      if (!old_slew || new_slew < *old_slew)
        set_slew(timing_mode, clock_edge, new_slew);
      break;
    default:
      LOG_FATAL << std::format("Unknown timing mode: {} for pin {}", *timing_mode, _name);
  }
}

void Pin::update_capacitance(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge)
{
  float_t total_capacitance{};
  for (const auto& arc : _fanout) {
    total_capacitance += arc->get_capacitance(timing_mode, clock_edge);
  }
  set_capacitance(timing_mode, clock_edge, total_capacitance);
}

void Pin::update_arrival_time(const EnumTimingMode timing_mode, const EnumClockEdge from_clock_edge, const EnumClockEdge to_clock_edge,
                              const float_t new_at, Arc* from_arc)
{
  LOG_ASSERT(from_arc->is_clock_edge_valid(from_clock_edge, to_clock_edge))
      << std::format("Invalid clock edge transition from {} to {} for arc {}", *from_clock_edge, *to_clock_edge, from_arc->get_name());
  const auto old_at{get_arrival_time(timing_mode, to_clock_edge)};
  switch (timing_mode) {
    case EnumTimingMode::MAX:
      if (!old_at || new_at > *old_at) {
        set_arrival_time(timing_mode, to_clock_edge, new_at);
        set_predecessor(timing_mode, from_clock_edge, to_clock_edge, from_arc);
      }
      break;
    case EnumTimingMode::MIN:
      if (!old_at || new_at < *old_at) {
        set_arrival_time(timing_mode, to_clock_edge, new_at);
        set_predecessor(timing_mode, from_clock_edge, to_clock_edge, from_arc);
      }
      break;
    default:
      LOG_FATAL << std::format("Unknown timing mode: {} for pin {}", *timing_mode, _name);
  }
}

void Pin::update_request_arrival_time(const EnumTimingMode timing_mode, EnumClockEdge, const EnumClockEdge to_clock_edge,
                                      const float_t new_rat)
{
  const auto old_rat{get_request_arrival_time(timing_mode, to_clock_edge)};
  switch (timing_mode) {
    case EnumTimingMode::MAX:
      if (!old_rat || new_rat > *old_rat) {
        set_request_arrival_time(timing_mode, to_clock_edge, new_rat);
      }
      break;
    case EnumTimingMode::MIN:
      if (!old_rat || new_rat < *old_rat) {
        set_request_arrival_time(timing_mode, to_clock_edge, new_rat);
      }
      break;
    default:
      LOG_FATAL << std::format("Unknown timing mode: {} for pin {}", *timing_mode, _name);
  }
}

void Pin::propagate_slew(const EnumTimingMode timing_mode, const EnumClockEdge from_clock_edge, const EnumClockEdge to_clock_edge) const
{
  for (auto& arc : _fanout) {
    arc->propagate_slew(timing_mode, from_clock_edge, to_clock_edge);
  }
}

void Pin::propagate_delay(const EnumTimingMode timing_mode, const EnumClockEdge from_clock_edge, const EnumClockEdge to_clock_edge) const
{
  for (const auto& arc : _fanout) {
    arc->propagate_delay(timing_mode, from_clock_edge, to_clock_edge);
  }
}

void Pin::propagate_arrival_time(const EnumTimingMode timing_mode, const EnumClockEdge from_clock_edge,
                                 const EnumClockEdge to_clock_edge) const
{
  for (const auto& arc : _fanout) {
    arc->propagate_arrival_time(timing_mode, from_clock_edge, to_clock_edge);
  }
}

void Pin::propagate_request_arrival_time(const EnumTimingMode timing_mode, const EnumClockEdge from_clock_edge,
                                         const EnumClockEdge to_clock_edge) const
{
  for (const auto& arc : _fanout) {
    arc->propagate_request_arrival_time(timing_mode, from_clock_edge, to_clock_edge);
  }
}

void Pin::set_slew(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge, const float_t slew)
{
  _slew[+timing_mode][+clock_edge] = slew;
}

void Pin::set_capacitance(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge, const float_t capacitance)
{
  _capacitance[+timing_mode][+clock_edge] = capacitance;
}

void Pin::set_arrival_time(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge, float_t arrival_time)
{
  _arrival_time[+timing_mode][+clock_edge] = arrival_time;
}

void Pin::set_request_arrival_time(const EnumTimingMode timing_mode, const EnumClockEdge clock_edge, float_t request_arrival_time)
{
  _request_arrival_time[+timing_mode][+clock_edge] = request_arrival_time;
}

void Pin::set_predecessor(const EnumTimingMode timing_mode, const EnumClockEdge from_clock_edge, const EnumClockEdge to_clock_edge,
                          Arc* predecessor)
{
  _predecessor[+timing_mode][+to_clock_edge] = std::pair{from_clock_edge, predecessor};
}

void Pin::reset_arrival_time(EnumTimingMode timing_mode, EnumClockEdge clock_edge)
{
  if (is_primary_input()) {
    _arrival_time[+timing_mode][+clock_edge] = 0;
    return;
  }
  _arrival_time[+timing_mode][+clock_edge].reset();
}

void Pin::reset_all_arrival_time()
{
  FOREACH_EL_RF([this](const auto el, const auto rf) { this->reset_arrival_time(el, rf); });
}

void Pin::connect_to(Net& net)
{
  switch (_type) {
    using enum EnumPinType;
    case INPUT:
    case PRIMARY_OUTPUT:
      net.add_sink(this);
      break;
    case OUTPUT:
    case PRIMARY_INPUT:
      net.set_source(this);
      break;
  }
}

std::optional<std::pair<EnumClockEdge, Arc*>> Pin::get_predecessor(EnumTimingMode timing_mode, EnumClockEdge clock_edge) const
{
  return _predecessor[+timing_mode][+clock_edge];
}

}  // namespace mySTA